libp2p-xtra/src/upgrade.rs

```diff
@@ -0,0 +1,125 @@
+use crate::verify_peer_id::VerifyPeerId;
+use crate::Connection;
+use futures::channel::mpsc;
+use futures::AsyncRead;
+use futures::AsyncWrite;
+use futures::FutureExt;
+use futures::SinkExt;
+use futures::StreamExt;
+use libp2p_core::identity::Keypair;
+use libp2p_core::transport::timeout::TransportTimeout;
+use libp2p_core::transport::Boxed;
+use libp2p_core::upgrade;
+use libp2p_core::upgrade::Version;
+use libp2p_core::Endpoint;
+use libp2p_core::Transport;
+use libp2p_noise as noise;
+use multistream_select::NegotiationError;
+use std::time::Duration;
+use void::Void;
+use yamux::Mode;
+
+/// Upgrades the given [`Transport`].
+///
+/// We apply:
+/// - Noise encryption and authentication
+/// - PeerID verification for each connection
+/// - Yamux multiplexing
+/// - Connection upgrade timeout
+pub fn transport<T>(
+    transport: T,
+    identity: &Keypair,
+    supported_inbound_protocols: Vec<&'static str>,
+    connection_timeout: Duration,
+) -> Boxed<Connection>
+where
+    T: Transport + Clone + Send + Sync + 'static,
+    T::Output: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+    T::Error: Send + Sync,
+    T::Listener: Send + 'static,
+    T::Dial: Send + 'static,
+    T::ListenerUpgrade: Send + 'static,
+{
+    let identity = noise::Keypair::<noise::X25519Spec>::new()
+        .into_authentic(identity)
+        .expect("ed25519 signing does not fail");
+
+    let authenticated = transport.and_then(|conn, endpoint| {
+        upgrade::apply(
+            conn,
+            noise::NoiseConfig::xx(identity).into_authenticated(),
+            endpoint,
+            Version::V1,
+        )
+    });
+
+    let peer_id_verified = VerifyPeerId::new(authenticated);
+
+    let multiplexed = peer_id_verified.and_then(|(peer_id, conn), endpoint| {
+        upgrade::apply(
+            conn,
+            upgrade::from_fn::<_, _, _, _, _, Void>(
+                b"/yamux/1.0.0",
+                move |conn, endpoint| async move {
+                    Ok(match endpoint {
+                        Endpoint::Dialer => (
+                            peer_id,
+                            yamux::Connection::new(conn, yamux::Config::default(), Mode::Client),
+                        ),
+                        Endpoint::Listener => (
+                            peer_id,
+                            yamux::Connection::new(conn, yamux::Config::default(), Mode::Server),
+                        ),
+                    })
+                },
+            ),
+            endpoint,
+            Version::V1,
+        )
+    });
+
+    let protocols_negotiated = multiplexed.map(move |(peer, mut connection), _| {
+        let control = connection.control();
+
+        let (mut sender, receiver) = mpsc::unbounded();
+
+        let worker = async move {
+            while let Ok(Some(stream)) = connection.next_stream().await {
+                let _ = sender.send(stream).await; // ignore error for now.
```

Comment:
❓ Why is this okay? Is this something that we're supposed to deal with at some point?

Reply:
I think we should actually:

a) use a bounded channel to apply back-pressure
b) break from this loop if we fail to send (which can only fail because the receiver is down)